### PR TITLE
mc-9607 Detect enumeration values

### PR DESCRIPTION
**Pull request by aaronforshaw**
Add generic detection of enumeration values.

Subclasses should override the new methods for counting and getting distinct values, and for choosing which data types to check as possible enumerations.

See corresponding pull requests for mdm-plugin-database-sqlserver and mdm-plugin-database-postgresql.

### CODE REVIEW

File: src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy [hunk 1]
```diff
@@ -109,6 +109,30 @@ abstract class DatabaseDataModelImporterProviderServiceParameters<K extends Data
         ))
     Boolean databaseSSL
 
+    @ImportParameterConfig(
+        displayName = 'Detect Enumerations',
+        description = 'Whether to treat columns with small numbers of unique values as enumerations',
+        order = 1,
+        optional = true,
+        group = @ImportGroupConfig(
+                name = 'Configuration',
```
Comment (olliefreeman) [inline src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy:118]: Can we change the group to "Database Import Details" as per "Database Name(s)"

File: src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy [hunk 2]
```diff
@@ -109,6 +109,30 @@ abstract class DatabaseDataModelImporterProviderServiceParameters<K extends Data
...
+    Boolean detectEnumerations = false
+
+    @ImportParameterConfig(
+        displayName = 'Maximum Enumerations',
+        description = 'The maximum number of unique values to be interpreted as a defined enumeration',
+        order = 2,
+        optional = true,
+        group = @ImportGroupConfig(
+                name = 'Configuration',
+                order = 2
```
Comment (olliefreeman) [inline src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy:131]: Can we change the group to "Database Import Details" as per "Database Name(s)"

**Review by olliefreeman — CHANGES_REQUESTED** [review]
Can we change the group please for the parameters. Also while you're at it i just noticed the group for the Suffix parameter is all on its own, can you udpate it to be part of the "Model" group which can be fdoun in the ModelImporterProviderServiceParameters.

As a side note, if creating new import param groups please pay attention to the ordering of the group and the inherited groups, this decides where they appear in the list of params and therefore in the UI. It may be worth looking at the UI with the plugin to see if its laid out as you like

**Comment by aaronforshaw** [comment]
@olliefreeman I have tweaked the grouping and ordering. On the UI, the parameters for the SQL Server plugin now look like this:
Screenshot 2021-08-17 at 13 34 48
Screenshot 2021-08-17 at 13 35 11

File: src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy (hunk 1 above)
Comment (aaronforshaw) [reply src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy]: Done

File: src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy (hunk 2 above)
Comment (aaronforshaw) [reply src/main/groovy/uk/ac/ox/softeng/maurodatamapper/plugins/database/DatabaseDataModelImporterProviderServiceParameters.groovy]: Done

**Review by olliefreeman — APPROVED** [review]

**Pull request merged**